### PR TITLE
Pipelines samples

### PR DESCRIPTION
**Pull request by mairaw**

### CODE REVIEW

File: snippets/csharp/pipelines/MyConnection.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        private PipeReader reader;
+
+        public MyConnection(PipeReader reader)
+        {
+            this.reader = reader;
+        }
+
+        public void Abort()
+        {
+            // Cancel the pending read so the the process loop ends without an exception
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/MyConnection.cs]:
```suggestion
            // Cancel the pending read so the the process loop ends without an exception.
```

File: snippets/csharp/pipelines/MyConnection.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+                while (true)
+                {
+                    ReadResult result = await reader.ReadAsync();
+                    ReadOnlySequence<byte> buffer = result.Buffer;
+
+                    try
+                    {
+                        if (result.IsCanceled)
+                        {
+                            // The read was canceled, we can quit without reading the existing data
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/MyConnection.cs]:
```suggestion
                            // The read was canceled, we can quit without reading the existing data.
```

File: snippets/csharp/pipelines/MyConnection.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+                            break;
+                        }
+
+                        // Process all messages from the buffer, modifying the input buffer on each iteration
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/MyConnection.cs]:
```suggestion
                        // Process all messages from the buffer, modifying the input buffer on each iteration.
```

File: snippets/csharp/pipelines/MyConnection.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+                        while (TryParseMessage(ref buffer, out Message message))
+                        {
+                            await ProcessMessageAsync(message);
+                        }
+
+                        // There's no more data to be processed
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/MyConnection.cs]:
```suggestion
                        // There's no more data to be processed.
```

File: snippets/csharp/pipelines/MyConnection.cs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+                        if (result.IsCompleted)
+                        {
+                            break;
+                        }
+                    }
+                    finally
+                    {
+                        // Since we're processing all messages in the buffer, we can use the remaining buffer's Start and End
+                        // position to determine consumed and examined
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/MyConnection.cs]:
```suggestion
                        // position to determine consumed and examined.
```

File: snippets/csharp/pipelines/MyConnection1.cs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+            try
+            {
+                while (true)
+                {
+                    ReadResult result = await reader.ReadAsync(cancellationToken);
+                    ReadOnlySequence<byte> buffer = result.Buffer;
+
+                    try
+                    {
+                        // Process all messages from the buffer, modifying the input buffer on each iteration
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/MyConnection1.cs]:
```suggestion
                        // Process all messages from the buffer, modifying the input buffer on each iteration.
```

File: snippets/csharp/pipelines/MyConnection1.cs [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+                        while (TryParseMessage(ref buffer, out Message message))
+                        {
+                            await ProcessMessageAsync(message);
+                        }
+
+                        // There's no more data to be processed
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/MyConnection1.cs]:
```suggestion
                        // There's no more data to be processed.
```

File: snippets/csharp/pipelines/MyConnection1.cs [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+                        if (result.IsCompleted)
+                        {
+                            if (buffer.Length > 0)
+                            {
+                                // We have an incomplete message and there's no more data to process
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/MyConnection1.cs]:
```suggestion
                                // We have an incomplete message and there's no more data to process.
```

File: snippets/csharp/pipelines/MyConnection1.cs [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+                                throw new InvalidDataException("Incomplete message!");
+                            }
+                            break;
+                        }
+                    }
+                    finally
+                    {
+                        // Since we're processing all messages in the buffer, we can use the remaining buffer's Start and End
+                        // position to determine consumed and examined
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/MyConnection1.cs]:
```suggestion
                        // position to determine consumed and examined.
```

File: snippets/csharp/pipelines/MyPipeWriter.cs [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+namespace Pipes
+{
+    class MyPipeWriter
+    {
+        // Reverted.  Recommend formatting for table   --------------------------------------
+        #region snippet
+        async Task WriteHelloAsync(PipeWriter writer, CancellationToken cancellationToken = default)
+        {
+            // Request at least 5 bytes from the PipeWriter
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/MyPipeWriter.cs]:
```suggestion
            // Request at least 5 bytes from the PipeWriter.
```

File: snippets/csharp/pipelines/MyPipeWriter.cs [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+            Memory<byte> memory = writer.GetMemory(5);
+
+            // Request at least 5 bytes from the PipeWriter
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/MyPipeWriter.cs]: duplicated comment ?

```suggestion
```

File: snippets/csharp/pipelines/MyPipeWriter.cs [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+            // Write directly into the buffer
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/MyPipeWriter.cs]:
```suggestion
            // Write directly into the buffer.
```

File: snippets/csharp/pipelines/MyPipeWriter.cs [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+            int written = Encoding.ASCII.GetBytes("Hello".AsSpan(), memory.Span);
+
+            // Tell the writer how many bytes we wrote
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/MyPipeWriter.cs]:
```suggestion
            // Tell the writer how many bytes we wrote.
```

File: snippets/csharp/pipelines/MyPipeWriter.cs [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+
+    class MyPipeWriter2
+    {
+        // Reverted.  Recommend formatting for table   --------------------------------------
+        #region snippet2
+        async Task WriteHelloAsync(PipeWriter writer, CancellationToken cancellationToken = default)
+        {
+            byte[] helloBytes = Encoding.ASCII.GetBytes("Hello");
+
+            // Write helloBytes to the writer, there's no need to call Advance here (Write does that)
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/MyPipeWriter.cs]:
```suggestion
            // Write helloBytes to the writer, there's no need to call Advance here (Write does that).
```

File: snippets/csharp/pipelines/Pipe.cs [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+            await Task.WhenAll(reading, writing);
+        }
+
+        async Task FillPipeAsync(Socket socket, PipeWriter writer)
+        {
+            const int minimumBufferSize = 512;
+
+            while (true)
+            {
+                // Allocate at least 512 bytes from the PipeWriter
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/Pipe.cs]:
```suggestion
                // Allocate at least 512 bytes from the PipeWriter.
```

File: snippets/csharp/pipelines/Pipe.cs [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+                Memory<byte> memory = writer.GetMemory(minimumBufferSize);
+                try
+                {
+                    int bytesRead = await socket.ReceiveAsync(memory, SocketFlags.None);
+                    if (bytesRead == 0)
+                    {
+                        break;
+                    }
+                    // Tell the PipeWriter how much was read from the Socket
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/Pipe.cs]:
```suggestion
                    // Tell the PipeWriter how much was read from the Socket.
```

File: snippets/csharp/pipelines/Pipe.cs [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+                    writer.Advance(bytesRead);
+                }
+                catch (Exception ex)
+                {
+                    LogError(ex);
+                    break;
+                }
+
+                // Make the data available to the PipeReader
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/Pipe.cs]:
```suggestion
                // Make the data available to the PipeReader.
```

File: snippets/csharp/pipelines/Pipe.cs [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+                FlushResult result = await writer.FlushAsync();
+
+                if (result.IsCompleted)
+                {
+                    break;
+                }
+            }
+
+            // Tell the PipeReader that there's no more data coming
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/Pipe.cs]:
```suggestion
            // Tell the PipeReader that there's no more data coming.
```

File: snippets/csharp/pipelines/Pipe.cs [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+        async Task ReadPipeAsync(PipeReader reader)
+        {
+            while (true)
+            {
+                ReadResult result = await reader.ReadAsync();
+                ReadOnlySequence<byte> buffer = result.Buffer;
+
+                while (TryReadLine(ref buffer, out ReadOnlySequence<byte> line))
+                {
+                    // Process the line
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/Pipe.cs]:
```suggestion
                    // Process the line.
```

File: snippets/csharp/pipelines/Pipe.cs [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+                    ProcessLine(line);
+                }
+
+                // Tell the PipeReader how much of the buffer we have consumed
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/Pipe.cs]:
```suggestion
                // Tell the PipeReader how much of the buffer we have consumed.
```

File: snippets/csharp/pipelines/Pipe.cs [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+                reader.AdvanceTo(buffer.Start, buffer.End);
+
+                // Stop reading if there's no more data coming
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/Pipe.cs]:
```suggestion
                // Stop reading if there's no more data coming.
```

File: snippets/csharp/pipelines/Pipe.cs [hunk 22]
```diff
@@ -0,0 +1,124 @@
...
+                if (result.IsCompleted)
+                {
+                    break;
+                }
+            }
+
+            // Mark the PipeReader as complete
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/Pipe.cs]:
```suggestion
            // Mark the PipeReader as complete.
```

File: snippets/csharp/pipelines/Pipe.cs [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+            reader.Complete();
+        }
+
+        bool TryReadLine(ref ReadOnlySequence<byte> buffer, out ReadOnlySequence<byte> line)
+        {
+            // Look for a EOL in the buffer
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/Pipe.cs]:
```suggestion
            // Look for a EOL in the buffer.
```

File: snippets/csharp/pipelines/Pipe.cs [hunk 24]
```diff
@@ -0,0 +1,124 @@
...
+            SequencePosition? position = buffer.PositionOf((byte)'\n');
+
+            if (position == null)
+            {
+                line = default;
+                return false;
+            }
+
+            // Skip the line + the \n
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/Pipe.cs]:
```suggestion
            // Skip the line + the \n.
```

File: snippets/csharp/pipelines/ProcessLinesAsync.cs [hunk 25]
```diff
@@ -0,0 +1,74 @@
...
+        #region snippet
+        async Task ProcessLinesAsync(NetworkStream stream)
+        {
+            byte[] buffer = ArrayPool<byte>.Shared.Rent(1024);
+            var bytesBuffered = 0;
+            var bytesConsumed = 0;
+
+            while (true)
+            {
+                // Calculate the amount of bytes remaining in the buffer
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/ProcessLinesAsync.cs]:
```suggestion
                // Calculate the amount of bytes remaining in the buffer.
```

File: snippets/csharp/pipelines/ProcessLinesAsync.cs [hunk 26]
```diff
@@ -0,0 +1,74 @@
...
+                var bytesRemaining = buffer.Length - bytesBuffered;
+
+                if (bytesRemaining == 0)
+                {
+                    // Double the buffer size and copy the previously buffered data into the new buffer
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/ProcessLinesAsync.cs]:
```suggestion
                    // Double the buffer size and copy the previously buffered data into the new buffer.
```

File: snippets/csharp/pipelines/ProcessLinesAsync.cs [hunk 27]
```diff
@@ -0,0 +1,74 @@
...
+                    var newBuffer = ArrayPool<byte>.Shared.Rent(buffer.Length * 2);
+                    Buffer.BlockCopy(buffer, 0, newBuffer, 0, buffer.Length);
+                    // Return the old buffer to the pool
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/ProcessLinesAsync.cs]:
```suggestion
                    // Return the old buffer to the pool.
```

File: snippets/csharp/pipelines/ProcessLinesAsync.cs [hunk 28]
```diff
@@ -0,0 +1,74 @@
...
+                }
+
+                var bytesRead = await stream.ReadAsync(buffer, bytesBuffered, bytesRemaining);
+                if (bytesRead == 0)
+                {
+                    // EOF
+                    break;
+                }
+
+                // Keep track of the amount of buffered bytes
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/ProcessLinesAsync.cs]:
```suggestion
                // Keep track of the amount of buffered bytes.
```

File: snippets/csharp/pipelines/ProcessLinesAsync.cs [hunk 29]
```diff
@@ -0,0 +1,74 @@
...
+                bytesBuffered += bytesRead;
+                var linePosition = -1;
+
+                do
+                {
+                    // Look for a EOL in the buffered data
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/ProcessLinesAsync.cs]:
```suggestion
                    // Look for a EOL in the buffered data.
```

File: snippets/csharp/pipelines/ProcessLinesAsync.cs [hunk 30]
```diff
@@ -0,0 +1,74 @@
...
+                    linePosition = Array.IndexOf(buffer, (byte)'\n', bytesConsumed, bytesBuffered - bytesConsumed);
+
+                    if (linePosition >= 0)
+                    {
+                        // Calculate the length of the line based on the offset
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/ProcessLinesAsync.cs]:
```suggestion
                        // Calculate the length of the line based on the offset.
```

File: snippets/csharp/pipelines/ProcessLinesAsync.cs [hunk 31]
```diff
@@ -0,0 +1,74 @@
...
+                        var lineLength = linePosition - bytesConsumed;
+
+                        // Process the line
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/ProcessLinesAsync.cs]:
```suggestion
                        // Process the line.
```

File: snippets/csharp/pipelines/ProcessLinesAsync.cs [hunk 32]
```diff
@@ -0,0 +1,74 @@
...
+                        ProcessLine(buffer, bytesConsumed, lineLength);
+
+                        // Move the bytesConsumed to skip past the line we consumed (including \n)
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/ProcessLinesAsync.cs]:
```suggestion
                        // Move the bytesConsumed to skip past the line we consumed (including \n).
```

File: snippets/csharp/pipelines/Program.cs [hunk 33]
```diff
@@ -0,0 +1,49 @@
...
+    class Program
+    {
+        // Reverted   - Recommend formatting code for Tables -no Horizontal bar end code here
+        #region snippet
+        public static void Main(string[] args)
+        {
+            var writeScheduler = new SingleThreadPipeScheduler();
+            var readScheduler = new SingleThreadPipeScheduler();
+
+            // Tell the Pipe what schedulers to use, we also disable the SynchronizationContext
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/Program.cs]:
```suggestion
            // Tell the Pipe what schedulers to use, we also disable the SynchronizationContext.
```

File: snippets/csharp/pipelines/Program.cs [hunk 34]
```diff
@@ -0,0 +1,49 @@
...
+        {
+            var writeScheduler = new SingleThreadPipeScheduler();
+            var readScheduler = new SingleThreadPipeScheduler();
+
+            // Tell the Pipe what schedulers to use, we also disable the SynchronizationContext 
+            var options = new PipeOptions(readerScheduler: readScheduler, writerScheduler: writeScheduler, useSynchronizationContext: false);
+            var pipe = new Pipe(options);
+        }
+
+        // This is a sample scheduler that async callbacks on a single dedicated thread
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/Program.cs]:
```suggestion
        // This is a sample scheduler that async callbacks on a single dedicated thread.
```

File: snippets/csharp/pipelines/ReadSingleMsg.cs [hunk 35]
```diff
@@ -0,0 +1,77 @@
...
+                // In the event that we don't parse any message successfully, mark consumed as nothing
+                // and examined as the entire buffer.
+                SequencePosition consumed = buffer.Start;
+                SequencePosition examined = buffer.End;
+
+                try
+                {
+                    if (TryParseMessage(ref buffer, out Message message))
+                    {
+                        // We successfully parsed a single message so mark the start as the parsed buffer as consumed
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/ReadSingleMsg.cs]:
```suggestion
                        // We successfully parsed a single message so mark the start as the parsed buffer as consumed.
```

File: snippets/csharp/pipelines/ReadSingleMsg.cs [hunk 36]
```diff
@@ -0,0 +1,77 @@
...
+                        // TryParseMessage trims the buffer to point to the data after the message was parsed
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/ReadSingleMsg.cs]:
```suggestion
                        // TryParseMessage trims the buffer to point to the data after the message was parsed.
```

File: snippets/csharp/pipelines/ReadSingleMsg.cs [hunk 37]
```diff
@@ -0,0 +1,77 @@
...
+                        consumed = buffer.Start;
+
+                        // Examined is marked the same as consumed here so that the next call to ReadSingleMessageAsync
+                        // will process the next message if there is one
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/ReadSingleMsg.cs]:
```suggestion
                        // will process the next message if there is one.
```

File: snippets/csharp/pipelines/ReadSingleMsg.cs [hunk 38]
```diff
@@ -0,0 +1,77 @@
...
+                        examined = consumed;
+
+                        return message;
+                    }
+
+                    // There's no more data to be processed
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/ReadSingleMsg.cs]:
```suggestion
                    // There's no more data to be processed.
```

File: snippets/csharp/pipelines/ReadSingleMsg.cs [hunk 39]
```diff
@@ -0,0 +1,77 @@
...
+                    if (result.IsCompleted)
+                    {
+                        if (buffer.Length > 0)
+                        {
+                            // We have an incomplete message and there's no more data to process
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/ReadSingleMsg.cs]:
```suggestion
                            // We have an incomplete message and there's no more data to process.
```

**Comment by mairaw** [comment]
Thanks @davidfowl and @Youssef1313. 

I've added some more changes and fixes. Good to merge?

File: snippets/csharp/pipelines/MyConnection1.cs [hunk 40]
```diff
@@ -0,0 +1,71 @@
...
+                            await ProcessMessageAsync(message);
+                        }
+
+                        // There's no more data to be processed.
+                        if (result.IsCompleted)
+                        {
+                            if (buffer.Length > 0)
+                            {
+                                // The message is incomplete and there's no more data to process.
+                                throw new InvalidDataException("Incomplete message!");
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/MyConnection1.cs]:
```suggestion
                                throw new InvalidDataException("Incomplete message.");
```

File: snippets/csharp/pipelines/ReadSingleMsg.cs [hunk 41]
```diff
@@ -0,0 +1,77 @@
...
+                        return message;
+                    }
+
+                    // There's no more data to be processed.
+                    if (result.IsCompleted)
+                    {
+                        if (buffer.Length > 0)
+                        {
+                            // The message is incomplete and there's no more data to process.
+                            throw new InvalidDataException("Incomplete message!");
```
Comment (Youssef1313) [inline snippets/csharp/pipelines/ReadSingleMsg.cs]:
```suggestion
                            throw new InvalidDataException("Incomplete message.");
```

**Pull request merged**